### ISSUE TITLE
refactor test/e2e/framework

### ISSUE THREAD

**Issue by pohly** [kubernetes/kubernetes#66649]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

The E2E framework has grown organically over time. As a result, it is both hard to maintain and close to impossible to use outside of Kubernetes itself.

**What you expected to happen**:

The goal is to make it possible that other projects can:
- import test/e2e/framework as a vendor dependency without pulling in support for cloud providers and any tests (because of the huge dependency chains that a complete import causes)
- import support for cloud providers and tests separately if desired
- add custom tests without modifying the framework
- create and run a Ginkgo suite

One potential user for this is the Kubernetes CSI project, which currently relies on tests in Kubernetes to test the out-of-tree sidecar containers. Testing the "canary" sidecar containers is done manually by developers. There is an on-going effort to unify all sidecar containers in one repo (https://docs.google.com/document/d/1xC0Cm11WTebT9m_9yZbfzud24GJRxlp55gdR5QaMHmM/edit), which then could also import and run the E2E testing in TravisCI.

Other users are CSI driver developers.

For a full description of this proposal including an implementation plan, see this Google doc: https://docs.google.com/document/d/1lPG264ZlwPbIuFeP9P0-c0vXdmCC7wzwaapf53BbXBA/edit?usp=sharing

**Comment by pohly** [comment]
My thoughts about Viper..

Pros:
- different, unrelated packages can set defaults and get values independently *if* we establish some naming scheme for keys such that keys don't clash - but the same also works with `flag`
- no need to register valid keys: just define some Go struct, then call `viper.Unmarshal(&mystruct)`

Cons:
- because keys don't need to be registered, typos in key names (in env vars or config files) are not detected; instead mistyped keys are silently ignored
- for the same reason, a help string can at best show the default values of keys, and only if a default
  has been set
- `viper.Unmarshal` is unusable for `viper.Sub()` (https://github.com/spf13/viper/issues/307), which makes it tricky to unmarshal a subset of the keys into a smaller struct inside tests; they would have to use their own viper instances, but then that breaks having a central config file, or use a workaround (*)

Perhaps I misunderstood something, but I don't find this very appealing.

Here's what I would do:
- define a naming scheme for command line flags and env variables
- only keep framework settings in TestContext, populate it with `flag` (i.e. remove Viper to avoid the extra dependency for people importing the framework)
- each test package and cloud provider manages its own context, using either `flag` or (if they want) their own Viper instance

(*) As workaround one can do this:
```
type SubContext struct {
	Bar int
}
var outer struct {
	Sub SubContext
}
var inner = &outer.Sub

func main() {
    viper.Unmarshal(&outer)
}
```

Then this JSON config properly populates `inner.Bar`:
```
{
    "Sub": {
        "Bar": 4
    }
}
```

**Comment by pohly** [comment]
One more thing to consider are auxiliary files. At the moment, they get built into the test binary via test/e2e/generated/bindata.go, which has to be generated anew when adding or removing tests. One of the goals for reusing storage tests for external drivers was that the tests can be used and extended via external scripts, without recompilation (https://docs.google.com/document/d/1I6Rn8X3-B0UbC6Q3PlBHMcbbr5kv5cyYQdBwdL4d3hY/edit?ts=5b3ac117#heading=h.816dbc1j453). Is that something that the framework should support, for example by looking up files in a filesystem before falling back to builtin bindata (if, and only if bindata is used by the test suite)?

Or should the use of bindata perhaps be removed entirely? This might be a problem for usages where currently only the compiled test program needs to be deployed on a remote system.

**Comment by ixdy** [comment]
I haven't worked on the e2e test frameworks in a while, so I don't have many strong opinions here.

I don't think the framework was ever intended to be used outside the kubernetes core repo, though some folks do apparently do this.

I believe we added the bindata dependency so that you only needed an `e2e.test` binary to run e2e tests.

It's probably out of scope for what you're suggesting here, since it would likely be pretty disruptive, but if we were going to fully rewrite the e2e test framework, I would suggest abandoning ginkgo.

**Comment by pohly** [comment]
The intention to use Viper instead of flags was introduced in be2e1f5d96487e765cb071da7ee6f90d47ff53f6. That commit has no explanation in the commit message why Viper is considered better than flags and no link to an issue about that work.

@jayunit100 do you remember more about that? Nothing has been done since that commit in 2016, so is there really still any interest in moving to Viper? If not, should the effort be abandoned and the Viper support be removed?

**Comment by pohly** [comment]
Let me propose something for the Viper topic:
- let's keep and *encourage* defining command line flags, because then a) they can be discovered without reading source code or documentation and b) accidental conflicts between different flags get detected by the `flag` package
- in addition, let's use Viper as additional method to set the same flags via a configuration file
- let's move non-framework flags out of the framework's TestContext without changing their name (except where it makes sense)
- let's make Viper an optional feature of the test/e2e/framework. test/e2e.go will use it, so in-tree test authors which only want to use Viper can do so. Out-of-tree test suites are not force to use it.

Here's code which implements the first two points and demonstrates the third one with a test using flags and a test not using it:
https://github.com/pohly/kubernetes/commits/e2e-refactor-2018-07-27

**Comment by timothysc** [comment]
I'd welcome a proposal doc that breaks apart the work into well-defined stages.  

wrt viper, the intent was to have a supported config file mechanism to run the tests.

**Comment by pohly** [comment]
"Timothy St. Clair" <notifications@github.com> writes:
> I'd welcome a proposal doc that breaks apart the work into
> well-defined stages.

A Google doc? I can create one. But as there seems to be no objections
against what I outlined so far, I'll also continue with prototyping in
parallel.

> wrt viper, the intent was to have a supported config file mechanism to
> run the tests.

That makes sense, and it is working now in my prototype branch for all
config options without changing the command line interface.

**Comment by jayunit100** [comment]
hi @pohly !  ... just saw this :  to answer your questions... 

- bindata was added because we needed to 'remember' the yaml files in things like 'examples/' tests for a particular release, otherwise, when trying to see if your e2e suite passed, you have to get the right corresponding static yaml files (there were many yaml files from the codebase as input to the tests, i.e. examples/ ones).  So, the binary now 'freezes' the static assets so your e2e run doesnt require any extra dependencies.
- the viper stuff was to move towards something that didnt drown us in flags and was dynamic/ easy to pass in config files for easily reproducible bug reports etc.    viper configs can automatically bind flags so they dont force a particular idiom on anyone.

**Comment by pohly** [comment]
@timothysc I have created this Google doc: https://docs.google.com/document/d/1lPG264ZlwPbIuFeP9P0-c0vXdmCC7wzwaapf53BbXBA/edit#

Let's move the discussion there.

**Comment by pohly** [comment]
Updated prototype, with cloud provider specific code in separate packages: https://github.com/pohly/kubernetes/commits/e2e-refactor-2018-07-31

It builds and runs tests in a local cluster, but needs further testing and I have some more ideas that I want to try out.

**Comment by BenTheElder** [comment]
/cc
These sound like great goal, will try to keep up with this 👍

**Comment by neolit123** [comment]
@jayunit100 
> the viper stuff was to move towards something that didnt drown us in flags and was dynamic/ easy to pass in config files for easily reproducible bug reports etc. viper configs can automatically bind flags so they dont force a particular idiom on anyone.

that very reasonable.

@pohly 
thank you for working on this.
just saw the sig-testing meeting. 

we also need something like viper in the context of kubeadm.
the number of issues and pending PRs is concerning for such a small project like viper. the project is lacking regular maintenance.

i think k8s pretty much needs a homebrew replacement as a library that does exactly the same thing - i.e. re-inventing the wheel. forking viper is also an option (probably a bad one).

> in addition, let's use Viper as additional method to set the same flags via a configuration file

this feels like an extra work for the e2e framework maintainers.
do you think it would be way too risky to completely switch to viper at this point?

> let's make Viper an optional feature of the test/e2e/framework. test/e2e.go will use it, so in-tree test authors which only want to use Viper can do so. Out-of-tree test suites are not force to use it.

wouldn't it be better to have the same feature set across the whole usage spectrum?
e.g. again allow viper for both.

**Comment by pohly** [comment]
"Lubomir I. Ivanov" <notifications@github.com> writes:
> we also need something like viper in the context of kubeadm.  the
> number of issues and pending PRs is concerning for such a small
> project like viper. the project is lacking regular maintenance.

Yes, I was also a bit concerned about that.

> i think k8s pretty much needs a homebrew replacement as a library that
> does exactly the same thing - i.e. re-inventing the wheel. forking
> viper is also an option (probably a bad one).

In some Viper PR comments it was said that others have already forked
Viper because they needed to host their fixes or enhancements. Perhaps
a maintenance team can be formed for it, either in its current location
or elsewhere?

>> in addition, let's use Viper as additional method to set the same flags via a configuration file
>
> this feels like an extra work for the e2e framework maintainers.
> do you think it would be way too risky to completely switch to viper
> at this point?

Using flags is automatic and therefore no extra work.
flag.String("my.new.config.option", ...) will also make it possible to
write a config file with that option:

my:
   new:
      config:
         option: "awesome"

The naming of the command line options is crucial here to get nice
hierarchical config files without conflicts between options defined
independently in different tests. I was trying to come up with a
recommended naming scheme in the commit enabling Viper for flags:
https://github.com/pohly/kubernetes/commit/1ebc471b914e24334793fd6c6f86a082a6bc73a1

In that proposal I chose to let a test file test/e2e/foo/context.go
define its flags as "foo.bar", mostly for the sake of brevity. I was
also considering to require some common prefix, but which one? "test"
conflicts with the test package. Perhaps "config"? Then "config.foo.bar"
could be mapped to a config entry "foo.bar" without that prefix.

I'm still a bit worried about usability: flags can be listed and thus
discovered, but not all flags necessarily can be set also in a config
file. For example, ginkgo.focus can't be used in a config file because
ginkgo doesn't use Viper. The same will be true for tests unless all
test authors call Viper to get their per-test config options with the
workaround from https://github.com/kubernetes/kubernetes/issues/66649#issuecomment-408108486

>> let's make Viper an optional feature of the
>> test/e2e/framework. test/e2e.go will use it, so in-tree test authors
>> which only want to use Viper can do so. Out-of-tree test suites are
>> not force to use it.
>
> wouldn't it be better to have the same feature set across the whole usage spectrum?
> e.g. again allow viper for both.

Viper is *allowed* to be used by all test suites, it's just not *forced*
upon them by the framework(*) - basically let the test suite maintainer
decide whether they need a config file and are willing to accept the
additional vendored code to get that feature.

(*) Tests that call Viper directly still create such a hard
dependency. I haven't thought about whether it would be worthwhile to
avoid that - probably yes. The framework could provide an abstract API
for populating a struct, and the implementation could map that to
command line flags and/or entries in a config file.

**Comment by pohly** [comment]
Vendoring test/e2e/framework into external projects is currently blocked by not publishing the cli-runtime repository under github.com/kubernetes (https://github.com/kubernetes/community/issues/2626 = https://github.com/kubernetes/org/issues/58).

**Comment by pohly** [comment]
Status update: https://github.com/pohly/kubernetes/commits/e2e-refactor-2018-09-04 can be imported and used in a different project, see https://github.com/pohly/csi-e2e for an example.

**Comment by neolit123** [comment]
@pohly 
the similar kubeadm issue is tracked here:
https://github.com/kubernetes/kubeadm/issues/1040

> In some Viper PR comments it was said that others have already forked
Viper because they needed to host their fixes or enhancements. Perhaps
a maintenance team can be formed for it, either in its current location
or elsewhere?

i'm all +1 for that.
to my understanding forking Viper and maintaining it ourself is not a big problem from the k8s perspective. but i have just asked in #sig-arch to see if there are some `no-no`s about this.

if OK, we can just replace `//vendor/github.com/spf13/viper` with another community maintained repo.

**Comment by neolit123** [comment]
> The naming of the command line options is crucial here to get nice
hierarchical config files without conflicts between options defined
independently in different tests. I was trying to come up with a
recommended naming scheme in the commit enabling Viper for flags:
pohly@1ebc471

i think i might be missing some context for this, due to my lack of understanding the e2e framework.
can you provide an example of how conflicts can happen? [1]

> In that proposal I chose to let a test file test/e2e/foo/context.go
define its flags as "foo.bar", mostly for the sake of brevity. I was
also considering to require some common prefix, but which one? "test"
conflicts with the test package. Perhaps "config"? Then "config.foo.bar"
could be mapped to a config entry "foo.bar" without that prefix.

is `test-e2e-config` available and/or does it make sense?

> I'm still a bit worried about usability: flags can be listed and thus
discovered, but not all flags necessarily can be set also in a config
file. For example, ginkgo.focus can't be used in a config file because
ginkgo doesn't use Viper. The same will be true for tests unless all
test authors call Viper to get their per-test config options with the
workaround from #66649 (comment)

i think we might be able to just list flags that are not part of the config next to flags that are part of the config, possibly separating them into two sections.

again missing some context as i do not understand the per-test config options.

**Comment by pohly** [comment]
"Lubomir I. Ivanov" <notifications@github.com> writes:

>  > The naming of the command line options is crucial here to get nice
> hierarchical config files without conflicts between options defined
> independently in different tests. I was trying to come up with a
> recommended naming scheme in the commit enabling Viper for flags:
> pohly@1ebc471
>
> i think i might be missing some context for this, due to my lack of understanding the e2e framework.
> can you provide an example of how conflicts can happen?

Right now, all flags and options are defined inside
test/e2e/framework/test_context.go. That way there can't be any
conflicts between different tests.

But defining config options for tests inside the framework makes it
impossible to add new or external tests without touching the
framework. A test author needs the ability to define and retrieve
options inside his test code.

That leads to potential conflicts when names overlap. For example,
suppose I import the in-tree framework and test/e2e/storage tests into a
custom E2E suite for my own CSI driver, like I do in
https://github.com/pohly/csi-e2e. Then I add a custom test to that suite
which uses a "-foobar" command line flag to configure some value. Now
someone adds that same "-foobar" flag in a patch for test/e2e/storage ->
when trying to update the source code, I have conflicting use of that
flag.

>> In that proposal I chose to let a test file test/e2e/foo/context.go
> define its flags as "foo.bar", mostly for the sake of brevity. I was
> also considering to require some common prefix, but which one? "test"
> conflicts with the test package. Perhaps "config"? Then "config.foo.bar"
> could be mapped to a config entry "foo.bar" without that prefix.
>
> is `test-e2e-config` available and/or does it make sense?

For use in a command line I find it too long. Right now, the only tests
which use Viper are test/e2e/instrumentation/monitoring/cadvisor.go and
test/e2e/instrumentation/logging/generic_soak.go, for example adding
"Cadvisor.MaxRetries".

Let's consider some possible ways to map that into the global namespace
for the command line flags:
  -test-e2e-config.cadvisor.maxretries
  -config.cadvisor.maxretries
  -instrumentation.cadvisor.maxretries (the naming I suggested in the
  commit)
  -config.instrumentation.cadvisor.maxretries (that naming plus
  a prefix that would tell users which flags can be specified also
  in a config file)

>> I'm still a bit worried about usability: flags can be listed and thus
> discovered, but not all flags necessarily can be set also in a config
> file. For example, ginkgo.focus can't be used in a config file because
> ginkgo doesn't use Viper. The same will be true for tests unless all
> test authors call Viper to get their per-test config options with the
> workaround from #66649 (comment)
>
> i think we might be able to just list flags that are not part of the
> config next to flags that are part of the config, possibly separating
> them into two sections.

Right now flags are handled by the normal "flag" module. There's no way
to distinguish between the two classes of flags automatically except via
some kind of naming convention.

Or were you thinking of writing documentation? I fear that this
documentation will be forever out-of-date. I'd rather have it as part of
the -help output and/or an option which writes a config file with the
default values, ideally with comments.

**Comment by neolit123** [comment]
On 6 September 2018 at 09:56, Patrick Ohly <notifications@github.com> wrote:
>
> "Lubomir I. Ivanov" <notifications@github.com> writes:
>
> > > The naming of the command line options is crucial here to get nice
> > hierarchical config files without conflicts between options defined
> > independently in different tests. I was trying to come up with a
> > recommended naming scheme in the commit enabling Viper for flags:
> > pohly@1ebc471
> >
> > i think i might be missing some context for this, due to my lack of understanding the e2e framework.
> > can you provide an example of how conflicts can happen?
>
> Right now, all flags and options are defined inside
> test/e2e/framework/test_context.go. That way there can't be any
> conflicts between different tests.
>
> But defining config options for tests inside the framework makes it
> impossible to add new or external tests without touching the
> framework. A test author needs the ability to define and retrieve
> options inside his test code.
>
> That leads to potential conflicts when names overlap. For example,
> suppose I import the in-tree framework and test/e2e/storage tests into a
> custom E2E suite for my own CSI driver, like I do in
> https://github.com/pohly/csi-e2e. Then I add a custom test to that suite
> which uses a "-foobar" command line flag to configure some value. Now
> someone adds that same "-foobar" flag in a patch for test/e2e/storage ->
> when trying to update the source code, I have conflicting use of that
> flag.
>

i think there is a bit of an underlying architectural problem here.
the framework does not have to know about the existence of individual
tests, but rather the framework should be used for individual tests.
if it's the other way around then this is more of a "library" not a
framework.
and then all tests should be added in a suite of tests based on the framework.

there should be:
- base structures defined in the framework that individual tests can
import and possibly extend.

if there is a single "suite" binary it should define:
- base flags that make sense for all tests.
- namespaces for custom flags that tests expose.

> >> In that proposal I chose to let a test file test/e2e/foo/context.go
> > define its flags as "foo.bar", mostly for the sake of brevity. I was
> > also considering to require some common prefix, but which one? "test"
> > conflicts with the test package. Perhaps "config"? Then "config.foo.bar"
> > could be mapped to a config entry "foo.bar" without that prefix.
> >
> > is `test-e2e-config` available and/or does it make sense?
>
> For use in a command line I find it too long. Right now, the only tests
> which use Viper are test/e2e/instrumentation/monitoring/cadvisor.go and
> test/e2e/instrumentation/logging/generic_soak.go, for example adding
> "Cadvisor.MaxRetries".
>
> Let's consider some possible ways to map that into the global namespace
> for the command line flags:
> -test-e2e-config.cadvisor.maxretries
> -config.cadvisor.maxretries
> -instrumentation.cadvisor.maxretries (the naming I suggested in the
> commit)
> -config.instrumentation.cadvisor.maxretries (that naming plus
> a prefix that would tell users which flags can be specified also
> in a config file)
>

it's fine as long as the prefixes would be unique in the long run.

> >> I'm still a bit worried about usability: flags can be listed and thus
> > discovered, but not all flags necessarily can be set also in a config
> > file. For example, ginkgo.focus can't be used in a config file because
> > ginkgo doesn't use Viper. The same will be true for tests unless all
> > test authors call Viper to get their per-test config options with the
> > workaround from #66649 (comment)
> >
> > i think we might be able to just list flags that are not part of the
> > config next to flags that are part of the config, possibly separating
> > them into two sections.
>
> Right now flags are handled by the normal "flag" module. There's no way
> to distinguish between the two classes of flags automatically except via
> some kind of naming convention.
>

that's not great, because there has to be a way to define what is
viper and what is a raw flag.
possibly moving everything to Viper would make it easier to user to
know that everything is config exposed.

but you did mention that some things are better as flags only.

> Or were you thinking of writing documentation? I fear that this
> documentation will be forever out-of-date. I'd rather have it as part of
> the -help output and/or an option which writes a config file with the
> default values, ideally with comments.
>

no, just talking about the `-help` output.

P.S. i've asked sig-architecture about forking of Viper and didn't get
a response yet...
might get one soon.

lubomir
--

**Comment by pohly** [comment]
"Lubomir I. Ivanov" <notifications@github.com> writes:
> On 6 September 2018 at 09:56, Patrick Ohly <notifications@github.com> wrote:
>> That leads to potential conflicts when names overlap. For example,
>> suppose I import the in-tree framework and test/e2e/storage tests into a
>> custom E2E suite for my own CSI driver, like I do in
>> https://github.com/pohly/csi-e2e. Then I add a custom test to that suite
>> which uses a "-foobar" command line flag to configure some value. Now
>> someone adds that same "-foobar" flag in a patch for test/e2e/storage ->
>> when trying to update the source code, I have conflicting use of that
>> flag.
>>
>
> i think there is a bit of an underlying architectural problem here.
> the framework does not have to know about the existence of individual
> tests, but rather the framework should be used for individual tests.

Exactly. Right now framework and tests are not separated like that,
which is what I am trying to fix here.

> if it's the other way around then this is more of a "library" not a
> framework.

I don't care whether it's called a framework or a library. The main
binary is defined by the test suite author, which calls functions in
test/e2e/framework, so I guess it's more like a library. OTOH, the test
suite has to be set up in a certain way, so that's more like extending a
framework.

> if there is a single "suite" binary it should define:
> - base flags that make sense for all tests.
> - namespaces for custom flags that tests expose.

Right. At the moment this is not cleanly separated, with (for example)
-volume-dir (a parameter that is only needed by the cluster tests) appearing
alongside -viper-config (a parameter handled by the framework).

>> Right now flags are handled by the normal "flag" module. There's no way
>> to distinguish between the two classes of flags automatically except via
>> some kind of naming convention.
>>
>
> that's not great, because there has to be a way to define what is
> viper and what is a raw flag.
> possibly moving everything to Viper would make it easier to user to
> know that everything is config exposed.

I think that was the original plan. Personally I don't like some aspects
of Viper, like not validating the config file (it silently ignores
unknown options, so typos are not reported) and not being able to list
the valid options. It also does not work well with options that are
decentralized (see
https://github.com/kubernetes/kubernetes/issues/66649#issuecomment-408108486
and the workaround that is necessary).

> but you did mention that some things are better as flags only.

What I meant is that some things can't be in a Viper config file because
the module which defines the flags doesn't use Viper. -test.v,
-ginkgo.focus and -alsologtostdout (from glog) fall into that camp.

>> Or were you thinking of writing documentation? I fear that this
>> documentation will be forever out-of-date. I'd rather have it as part of
>> the -help output and/or an option which writes a config file with the
>> default values, ideally with comments.
>>
>
> no, just talking about the `-help` output.

Then a Viper-only solution will not be very... helpful (see above
regarding not listing valid options). Sorry, bad pun, but I couldn't
resist ;-}

> P.S. i've asked sig-architecture about forking of Viper and didn't get
> a response yet...
> might get one soon.

I'm coming to the conclusion that Viper is overkill for just reading
some config file. I've started to write something simpler, let's see how
that goes.

**Comment by pohly** [comment]
"Lubomir I. Ivanov" <notifications@github.com> writes:
> On 6 September 2018 at 19:02, Patrick Ohly <notifications@github.com> wrote:
>> What I meant is that some things can't be in a Viper config file because
>> the module which defines the flags doesn't use Viper. -test.v,
>> -ginkgo.focus and -alsologtostdout (from glog) fall into that camp.
>>
>
> glog is something that we are trying to replace.

That would make sense. But then Ginkgo and the test module still remain.

> perhaps the architectural changes can address the issues with what
> flags the modules expose.

It might be possible to import flag definitions from the standard flag
package into some other package and the never call flag.Parse. pflags
supported that.

>> I'm coming to the conclusion that Viper is overkill for just reading
>> some config file. I've started to write something simpler, let's see how
>> that goes.
>>
>
> the kubelet did a homebrew solution and the response was that they
> regret it. it also introduced CLI reference doc generation problems.
> in kubeadm we also started doing that and the initial PRs got
> rejected, because it introduces a lot of boiler plate.
> kube-proxy also needs something like that because currently config
> values overrides CLI flags which is not very intuitive and inverse.

Has someone looked at https://github.com/jessevdk/go-flags? It seems to
do several things very well. It can generate docs in man format and it
has support for ini-style config files. I'm pretty sure adding support
for other output formats and hierarchical config files (ini is limited
to a flat list of groups) could be added.

But it might have the same issue as other utility packages (like
for example Viper): a single maintainer who has to keep up with a
growing number of external projects all coming with different requests.

**Comment by neolit123** [comment]
On 6 September 2018 at 21:43, Patrick Ohly <notifications@github.com> wrote:
>
> "Lubomir I. Ivanov" <notifications@github.com> writes:
> > On 6 September 2018 at 19:02, Patrick Ohly <notifications@github.com> wrote:
> >> What I meant is that some things can't be in a Viper config file because
> >> the module which defines the flags doesn't use Viper. -test.v,
> >> -ginkgo.focus and -alsologtostdout (from glog) fall into that camp.
> >>
> >
> > glog is something that we are trying to replace.
>
> That would make sense. But then Ginkgo and the test module still remain.
>
> > perhaps the architectural changes can address the issues with what
> > flags the modules expose.
>
> It might be possible to import flag definitions from the standard flag
> package into some other package and the never call flag.Parse. pflags
> supported that.
>

this might be a good option.

> >> I'm coming to the conclusion that Viper is overkill for just reading
> >> some config file. I've started to write something simpler, let's see how
> >> that goes.
> >>
> >
> > the kubelet did a homebrew solution and the response was that they
> > regret it. it also introduced CLI reference doc generation problems.
> > in kubeadm we also started doing that and the initial PRs got
> > rejected, because it introduces a lot of boiler plate.
> > kube-proxy also needs something like that because currently config
> > values overrides CLI flags which is not very intuitive and inverse.
>
> Has someone looked at https://github.com/jessevdk/go-flags? It seems to
> do several things very well. It can generate docs in man format and it
> has support for ini-style config files. I'm pretty sure adding support
> for other output formats and hierarchical config files (ini is limited
> to a flat list of groups) could be added.
>
> But it might have the same issue as other utility packages (like
> for example Viper): a single maintainer who has to keep up with a
> growing number of external projects all coming with different requests.
>

i haven't tried:
https://github.com/jessevdk/go-flags

we had a chat ealier with @BenTheElder and the idea of having
something maintained by the k8s team made a lot more sense.
i don't think we would be able to get what we want from other
libraries for this.

lubomir
--

**Comment by pohly** [comment]
I had another look at my own code - it's been too long already! ;-)

What I had forgotten is that a value from a Viper config will already update the same value as the flag package, thanks to `viperUnmarshall` in https://github.com/pohly/kubernetes/commit/f7cb59474eb187b6d3fcfcc090c83eb292590ae8#diff-e00287b5331f6d39959423afadb6b368R473.

So the situation now is that all tests that define their own options via flags already work well both with command line flags and with a Viper config, without the test depending on Viper. IMHO this approach together with a good naming convention for command line flags should be the suggested way for test authors to add configuration for their tests. I can even make Viper optional, to address my own concern about pulling it into a test suite when it isn't needed.

**Comment by pohly** [comment]
@neolit123 I've pushed another revision of my patch set: https://github.com/pohly/kubernetes/tree/e2e-refactor-2018-09-07. It's not tested much yet beyond ensuring that everything still builds.

The simpler API for defining a set of configuration options is here:
https://github.com/pohly/kubernetes/commit/69bbbbf3597a257814d8f363cc4665277516d14f#diff-58fcfea83a0fe45a637ccd0caeeb2788

I did not implement my own config parsing. Instead that code is just some helper to simplify the definition of normal command line flags. A configuration file is still read with Viper, but that now is a minor implementation detail as far as test authors are concerned. It could also become something else. I also made this Viper support optional, i.e. the framework can be used without pulling in Viper.

When Kubernetes decides how to do configuration files, adapting to that in test/e2e should be easy. Are there tracking issues for that work?

**Comment by neolit123** [comment]
@pohly 

sorry for not replying earlier, busy looking at bugs and tests.

> I did not implement my own config parsing. Instead that code is just some helper to simplify the definition of normal command line flags. A configuration file is still read with Viper, but that now is a minor implementation detail as far as test authors are concerned. It could also become something else. I also made this Viper support optional, i.e. the framework can be used without pulling in Viper.

i might get some more free time after 1.12 is released and do a mockup replacement for Viper. right now i still lack some understanding of the demands of the e2e framework, but i do understand what kubeadm needs.

BTW, the idea of homebrewing a replacement for viper did get some +1s in the sig-architecture slack chat instead of forking.

> When Kubernetes decides how to do configuration files, adapting to that in test/e2e should be easy. Are there tracking issues for that work?

here is the tracking issue for kubeadm:
https://github.com/kubernetes/kubeadm/issues/1040

but we haven't progressed much with decisions there.

**Comment by pohly** [comment]
One data point: not using the Viper-config file support in https://github.com/pohly/csi-e2e made it possible to drop also spf13/jwalterweatherman, spf13/afero, pelletier/go-toml, magiconair/properties, hashicorp/hcl, fsnotify.
